subsys/net/ip/tcp2.c

```diff
@@ -1761,11 +1761,12 @@ static void tcp_in(struct tcp *conn, struct net_pkt *pkt)
 				}
 				conn_ack(conn, + len);
 			}
-			k_sem_give(&conn->connect_sem);
+
 			next = TCP_ESTABLISHED;
```

Comment:
Indeed this change seems to fix the issue in the TLS tests, but I'm wondering, the actual TCP state is updated at line 1933, shouldn't the semaphore be fed after that? In case a higher priority thread takes precendence after feeding the semaphore, it might still fail at `net_tcp_queue_data` which verifies the TCP internal state.

Reply:
Hmm, that is a good point. Indeed the application might start to send more data even if the TCP has not yet fully finished the connection establishment. I will rework the PR more.